client/app/components/dashboards/AddWidgetDialog.jsx

```diff
@@ -0,0 +1,448 @@
+    const alreadyExists = includes(existingParamNames, mapping.mapTo);
+    return (
+      <div key={mapping.name} className={'row' + (isFirst ? '' : ' m-t-15')}>
+        <div className="col-xs-5">
+          <div className="form-control-static">{'{{ ' + mapping.name + ' }}'}</div>
+        </div>
+        <div className="col-xs-7">
+          <div>
+            <select
+              className="form-control"
+              value={mapping.type}
+              onChange={event => this.updateParamMapping(mapping, { type: event.target.value })}
+            >
+              <option value={MappingType.DashboardAddNew}>Add the parameter to the dashboard</option>
+              {
+                (existingParamNames.length > 0) &&
+                <option value={MappingType.DashboardMapToExisting}>Map to existing parameter</option>
+              }
+              <option value={MappingType.StaticValue}>Use static value for the parameter</option>
+              <option value={MappingType.WidgetLevel}>Keep the parameter at the widget level</option>
+            </select>
+          </div>
+          {
+            (mapping.type === MappingType.DashboardAddNew) &&
+            <div className={'m-t-10' + (alreadyExists ? ' has-error' : '')}>
+              <input
+                type="text"
+                className="form-control"
+                value={mapping.mapTo}
+                onChange={event => this.updateParamMapping(mapping, { mapTo: event.target.value })}
+              />
+              { alreadyExists &&
+                <div className="help-block">
+                  Dashboard parameter with this name already exists
+                </div>
+              }
+            </div>
+          }
+          {
+            (mapping.type === MappingType.DashboardMapToExisting) &&
+            <div className="m-t-10">
+              <select
+                className="form-control"
+                value={mapping.mapTo}
+                onChange={event => this.updateParamMapping(mapping, { mapTo: event.target.value })}
+                disabled={existingParamNames.length === 0}
+              >
+                {map(existingParamNames, name => (
+                  <option value={name} key={name}>{ name }</option>
+                ))}
+              </select>
+            </div>
+          }
+          {
+            (mapping.type === MappingType.StaticValue) &&
+            <div className="m-t-10">
+              <ParameterValueInput
+                type={mapping.param.type}
+                value={mapping.value}
+                enumOptions={mapping.param.enumOptions}
+                queryId={mapping.param.queryId}
+                onSelect={value => this.updateParamMapping(mapping, { value })}
+                clientConfig={this.props.clientConfig} // eslint-disable-line react/prop-types
+                Query={this.props.Query} // eslint-disable-line react/prop-types
+              />
+            </div>
+          }
```

Comment:
Same here as with the different parameter input types: let's use multiple render functions.

Reply:
Also it should be its own component.